File: python/wsfl/watermark.py
from PIL import Image, ImageDraw
import io
from timeit import timeit
from multiprocessing import Pool
from concurrent import futures

def watermark(_):
    front = Image.open('../../data/lena512color.jpg')
    stamp = Image.open('../../data/stamp.png')

    d = ImageDraw.Draw(front)
    d.bitmap((0,0), stamp.resize(front.size))
    bytes = io.BytesIO()
    front.save(bytes, 'JPEG')
    unuse = bytes.read()

N = 1000
@timeit
def synchronize():
    for _ in range(N):
        watermark(_)
    return N

@timeit
def process():
    with Pool(4) as p:
        p.map(watermark, range(N))
    return N

@timeit
def thread():
    with futures.ThreadPoolExecutor(4) as p:
        p.map(watermark, range(N))
    return N

synchronize()
process()
thread()

File: java/wsfl/src/main/java/wsfl/Watermark.java
package wsfl;

import java.awt.Graphics2D;
import java.awt.image.BufferedImage;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.IOException;
import java.util.stream.IntStream;

import javax.imageio.ImageIO;

public class Watermark {
	
	private int count = 1000;
	
	private void doWatermark() throws IOException {
		
		BufferedImage front = ImageIO.read(new File("../../data/lena512color.jpg"));
		BufferedImage stamp = ImageIO.read(new File("../../data/stamp.png"));
		
		BufferedImage stampResized = new BufferedImage(front.getWidth(), front.getHeight(), stamp.getType());
		Graphics2D gResized = stampResized.createGraphics();
		gResized.drawImage(stamp, 0, 0, front.getWidth(), front.getHeight(), null);
		gResized.dispose();
		
//		ImageIO.write(stampResized, "PNG", new File("tmp.png"));
		
		Graphics2D g = front.createGraphics();
		g.drawImage(stampResized, 0, 0, null);
		g.dispose();
		
		ByteArrayOutputStream out = new ByteArrayOutputStream();
		
		ImageIO.write(front, "JPEG", out);
		
		byte[] unuse = out.toByteArray();
	}
	
	public void synchronize(Timeit ti) {
		IntStream.range(0, count).forEach(i -> { 
			try {
				this.doWatermark();
			} catch (IOException e) {
				e.printStackTrace();
			} 
		} );
		ti.set("顺序", count);
	}
	
	public void multipleThread(Timeit ti) {
		IntStream.range(0, count).parallel().forEach(i -> { 
			try {
				this.doWatermark();
			} catch (IOException e) {
				e.printStackTrace();
			} 
		} );
		ti.set("多线程", count);
	}

	public static void main(String[] args) throws Exception {
		
		Watermark h = new Watermark();
		
		Timeit.timeit(h::synchronize);
		Timeit.timeit(h::multipleThread);
		
	}

}
